### DISCUSSION TITLE
support windows

### DISCUSSION THREAD

**Discussion by ChinaGuCheng** [danielpaulus/quicktime_video_hack#130] [Ideas]
Using libusb-win32 api and libusb0 driver can support windows, I have some experience

**Comment by danielpaulus** [comment]
Nice! I am trying it to today to see if I can get it to work as well. Thanks for helping me out here @ChinaGuCheng  :-)

**Reply by danielpaulus** [reply]
@ChinaGuCheng  
1. Do you know how to setup pkg-config with libusb-win32 correctly for this to work?

2. I see that libusb-win32 is in maintenance mode, did you also get it to work with the latest windows driver from the official project libusb.info ?

**Comment by ChinaGuCheng** [comment]
@danielpaulus 
1,
I don't use pkg-config, I download the libusb-win32 code and then import .h .lib .dll, download address: https://sourceforge.net/projects/libusb-win32/files/libusb-win32-releases/1.2.6.0/

2,
Yes, libusb is in maintenance mode, but it still works
I have tried the new libusb1.0, the function usb_set_configure will not succeed, but libusb-win32 can.

**Comment by danielpaulus** [comment]
I think I can just write a little C tool that sets the config using win32 libusb. Then bundle it with qvh and use it for setting the config.  
Found your issue: https://github.com/libusb/libusb/issues/1103

**Reply by ChinaGuCheng** [reply]
> I think I can just write a little C tool that sets the config using win32 libusb. Then bundle it with qvh and use it for setting the config. Found your issue: [libusb/libusb#1103](https://github.com/libusb/libusb/issues/1103)



This one is libusb1.0, you should use libusb-win32

libusb1.0 is cross-platform, supports libusb0 driver, libusbK driver, winusb driver, usbdk driver
https://github.com/libusb/libusb


libusb-win32 only supports Windows system, only supports libusb0 driver
https://sourceforge.net/projects/libusb-win32/files/libusb-win32-releases/1.2.6.0/

**Reply by danielpaulus** [reply]
@ChinaGuCheng  I Sent you an email just now, better remove your email address from the thread so you don't get spam emails.

**Reply by ChinaGuCheng** [reply]
> @ChinaGuCheng I Sent you an email just now, better remove your email address from the thread so you don't get spam emails.

Ok, I have sent the demo email to you.

**Comment by danielpaulus** [comment]
I was able to activate a device as well. 
Install the driver with infwizard, and ran `device_activator UID` then the device showed up with the additional config. Pretty cool.
I hope that libusb-1.0 will work with the libusb32 driver without further hassle, otherwise I have to add quite a bit of code for libusb32 compatibility. 🤞 

TL;DR It definitely works, the question is how much work is needed to get it into a useable state. 
My progress is on the `windows/support` branch.

**Comment by ChinaGuCheng** [comment]
libusb-win32 uses usb_set_configuration, and other operations can be done with libusb-1.0.

**Comment by ChinaGuCheng** [comment]
Here is the source code snippet for usbmuxd which uses libusb-win32 set configure on windows and libusb1.0 on other platforms.

**
#ifdef WIN32
		char serial[128] = { 0 };
		if ((res = libusb_get_string_descriptor_ascii(handle, devdesc.iSerialNumber, (uint8_t *)serial, 40)) <= 0) {
			usbmuxd_log(LL_WARNING, "libusb_get_string_descriptor_ascii error,  for device %d-%d: %d %s", bus, address, res, libusb_strerror(res));
			libusb_close(handle);
			return -1;
		}
		printf("serial [%s]\n", serial);

		//strcpy(serial, "c80a85d400342a3d730ac3dfe2daa05b5ce7d145");
		usb_win32_set_configuration(serial, desired_config);
		printf("usb_win32_set_configuration [%s][%d]\n", serial, desired_config);
		

		// Because the change was done via libusb-win32, we need to refresh the device on libusb;
		// otherwise, it will not pick up the new configuration and endpoints.
		// For now, let the next loop do this for us.
		libusb_close(handle);	
		return -2;
#else
		if((res = libusb_set_configuration(handle, desired_config)) != 0) {
			usbmuxd_log(LL_WARNING, "Could not set configuration %d for device %d-%d: %s", desired_config, bus, address, libusb_error_name(res));
			libusb_close(handle);
			return -1;
		}
		usbmuxd_log(LL_WARNING, "success set configuration %d for device %d-%d", desired_config, bus, address);
#endif
**
